daemon/migrator.c

```diff
@@ -0,0 +1,804 @@
+        approval_dtor(self);
+        g_free(self);
+    }
+}
+
+void
+approval_delete_cb(void *self)
+{
+    approval_delete(self);
+}
+
+/* ------------------------------------------------------------------------- *
+ * APPROVAL_ATTRIBUTES
+ * ------------------------------------------------------------------------- */
+
+static bool
+approval_valid(const approval_t *self)
+{
+    return approval_uid(self) != MIGRATION_UID_UNDEFINED;
+}
+
+static const gchar *
+approval_profile(const approval_t *self)
+{
+    /* Application profile path, i.e. application's desktop entry file path */
+    return self->apr_profile;
+}
+
+static uid_t
+approval_uid(const approval_t *self)
+{
+    /* UID from approval file path */
+    return self->apr_uid;
+}
+
+static stringset_t *
+approval_permissions(const approval_t *self)
+{
+    /* Approved permissions */
+    return self->apr_permissions;
+}
+
+static const gchar *
+approval_organization(const approval_t *self)
+{
+    /* The same as OrganizationName in Sailjail section */
+    return self->apr_organization;
+}
+
+static const gchar *
+approval_application(const approval_t *self)
+{
+    /* The same as ApplicationName in Sailjail section */
+    return self->apr_application;
+}
+
+#if DEAD_CODE
+static const gchar *
+approval_exec(const approval_t *self)
+{
+    /* Binary path, does not contain arguments */
+    return self->apr_exec;
+}
+#endif
+
+static const migrator_t *
+approval_migrator(const approval_t *self)
+{
+    return self->apr_migrator;
+}
+
+static const control_t *
+approval_control(const approval_t *self)
+{
+    return migrator_control(approval_migrator(self));
+}
+
+/* ------------------------------------------------------------------------- *
+ * UTILITY
+ * ------------------------------------------------------------------------- */
+
+#if DEAD_CODE
+static bool
+exec_matches_binary(const gchar *exec, const gchar *binary_path)
+{
+    gchar **argv = NULL;
+    gint    argc = 0;
+    bool    rv   = false;
+
+    /* Note: This can deal only with Exec lines that contain the
+     *       application as the first argument, i.e. it can't deal with
+     *       Exec lines that contains sailjail or invoker first.
+     */
+    if( g_shell_parse_argv(exec, &argc, &argv, NULL) &&
+        argc >= 1 && 
+        !g_strcmp0(argv[0], binary_path) )
+        rv = true;
+
+    g_strfreev(argv);
+    return rv;
+}
+#endif
+
+static void
+remove_approval_path(const gchar *path)
+{
+    /* Remove path and any empty parent directories until HOMESCREEN_DATA_PATH */
+    if( unlink(path) == 0 ) {
+        gchar *dir_path = path_dirname(path);
+        while( may_remove_approval_path(dir_path) && rmdir(dir_path) == 0 ) {
+            gchar *temp = dir_path;
+            dir_path = path_dirname(temp);
+            g_free(temp);
+        }
+        log_debug("%s cleaned up to %s", path, dir_path);
+        g_free(dir_path);
+    }
+    else {
+        log_err("%s could not remove: %m", path);
```

Comment:
nit / matter of opinion: I'd prefer to see error handling first. Basically so that when one is reading code it is immediately obvious that error case is handled and it can then be forgotten about vs. doing the other way around and one ends up with stack of error handlers that are far away from where it happened. In this case the situation is short and shallow so it does not really matter...